File: app/src/main/java/cn/guluwa/gulumusic/service/MusicAutoService.java
package cn.guluwa.gulumusic.service;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.support.v4.content.LocalBroadcastManager;
import android.util.Log;

import cn.guluwa.gulumusic.manage.AppManager;

/**
 * Created by guluwa on 2018/1/26.
 */

public class MusicAutoService extends Service {

    public static final String TAG = "MusicAutoService";

    /**
     * 播放器焦点管理器
     */
    private AudioFocusManager mAudioFocusManager;


    public MusicAutoService() {
    }

    //  通过 Binder 来保持 Activity 和 Service 的通信
    public MusicBinder binder = new MusicBinder(this);

    @Override
    public boolean onUnbind(Intent intent) {
        return super.onUnbind(intent);
    }

    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        mAudioFocusManager = new AudioFocusManager(this);
        Log.w(TAG, "MusicAutoService in onCreate");
    }

    @Override
    public void onDestroy() {
        binder.unbindProgressQuery();
        mAudioFocusManager.abandonAudioFocus();
        binder.getMediaPlayer().reset();
        binder.getMediaPlayer().release();
        binder.setMediaPlayer(null);
        AppManager.Companion.getInstance().setMusicAutoService(null);
        super.onDestroy();
        Log.w(TAG, "MusicAutoService in onDestroy");
    }

    /**
     * 主动结束服务
     */
    public void quit() {
        stopSelf();
    }

    public AudioFocusManager getAudioFocusManager() {
        return mAudioFocusManager;
    }
}

File: app/src/main/java/cn/guluwa/gulumusic/service/MusicAutoService.kt
package cn.guluwa.gulumusic.service

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.util.Log

import cn.guluwa.gulumusic.manage.AppManager
import cn.guluwa.gulumusic.manage.Contacts
import cn.guluwa.gulumusic.service.notification.MyNotificationManager
import cn.guluwa.gulumusic.utils.AppUtils
import com.google.gson.Gson


/**
 * Created by guluwa on 2018/1/26.
 */

class MusicAutoService : Service() {

    /**
     * 播放器焦点管理器
     */
    var audioFocusManager: AudioFocusManager? = null
        private set

    //  通过 Binder 来保持 Activity 和 Service 的通信
    var binder = MusicBinder(this)

    override fun onBind(intent: Intent): IBinder? {
        return binder
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        return Service.START_STICKY
    }

    override fun onCreate() {
        super.onCreate()
        audioFocusManager = AudioFocusManager(this)
    }

    fun initNotification() {
        MyNotificationManager.getInstance().showNotification(this)
    }

    private fun destroyNotification() {
        MyNotificationManager.getInstance().dismissNotification()
    }

    override fun onDestroy() {
        binder.unbindProgressQuery()
        audioFocusManager!!.abandonAudioFocus()
        binder.mediaPlayer!!.reset()
        binder.mediaPlayer!!.release()
        binder.mediaPlayer = null
        AppManager.getInstance().musicAutoService = null
        destroyNotification()
        super.onDestroy()
        Log.w(TAG, "MusicAutoService in onDestroy")
    }

    /**
     * 主动结束服务
     */
    fun quit() {
        println("MusicAutoService quit")
        AppUtils.setString("mCurrentSong", Gson().toJson(AppManager.getInstance().musicAutoService!!.binder.currentSong))
        AppUtils.setInteger(Contacts.PLAY_MODE, AppManager.getInstance().playMode)
        AppUtils.setString(Contacts.PLAY_STATUS, AppManager.getInstance().playStatus)
        stopSelf()
    }

    companion object {

        const val TAG = "MusicAutoService"
    }
}
